source/agora/consensus/Validation.d

```diff
@@ -223,38 +224,225 @@ unittest
+
+    /// no matching utxo => fail
+    tx_map.clear();
+    assert(!block.isValid(gen_block.header.height, gen_hash, findOutput));
+
+    tx_map[gen_tx.hashFull()] = [gen_tx];
+    assert(block.isValid(gen_block.header.height, gen_hash, findOutput));
+
+    tx_map.clear();  // genesis is spent
+    auto prev_txs = txs;
+    prev_txs.each!(tx => tx_map[tx.hashFull()] = [tx]);  // these will be spent
+
+    auto prev_block = block;
+    txs = makeChainedTransactions(gen_key, prev_txs, 1);
+    block = makeNewBlock(prev_block, txs);
+    assert(block.isValid(prev_block.header.height, prev_block.header.hashFull(),
+        findOutput));
+
+    assert(prev_txs.length > 0);  // sanity check
+    foreach (tx; prev_txs)
+    {
+        // one utxo missing from the set => fail
+        tx_map.remove(tx.hashFull);
+        assert(!block.isValid(prev_block.header.height, prev_block.header.hashFull(),
+            findOutput));
+
+        tx_map[tx.hashFull] = [tx];
+        assert(block.isValid(prev_block.header.height, prev_block.header.hashFull(),
+            findOutput));
+    }
+
+    // the key is hashMulti(hash(prev_tx), index)
+    Output[Hash] utxo_set;
+
+    foreach (idx, output; gen_tx.outputs)
+        utxo_set[hashMulti(gen_tx.hashFull, idx)] = output;
+
+    assert(utxo_set.length != 0);
+    const utxo_set_len = utxo_set.length;
+
+    // contains the used set of UTXOs during validation (to prevent double-spend)
+    Output[Hash] used_set;
+    UtxoFinder findNonSpent = (Hash hash, size_t index)
+    {
+        auto utxo_hash = hashMulti(hash, index);
+
+        if (utxo_hash in used_set)
+            return null;  // double-spend
+
+        if (auto utxo = utxo_hash in utxo_set)
+        {
+            used_set[utxo_hash] = *utxo;
+            return utxo;
+        }
+
+        return null;
+    };
+
+    // consumed all utxo => fail
+    txs = makeChainedTransactions(gen_key, null, 1);
+    block = makeNewBlock(gen_block, txs);
+    assert(block.isValid(gen_block.header.height, gen_block.header.hashFull(),
+            findNonSpent));
+
+    assert(used_set.length == utxo_set_len);  // consumed all utxos
+
+    // reset state
+    used_set.clear();
+
+    // consumed same utxo twice => fail
+    txs[$ - 1] = txs[$ - 2];
+    block = makeNewBlock(gen_block, txs);
+    assert(!block.isValid(gen_block.header.height, gen_block.header.hashFull(),
+            findNonSpent));
+
+    // we stopped validation due to a double-spend
+    assert(used_set.length == txs.length - 1);
+
+    txs = makeChainedTransactions(gen_key, prev_txs, 1);
+    block = makeNewBlock(gen_block, txs);
+    assert(block.isValid(gen_block.header.height, gen_block.header.hashFull(),
+        findOutput));
+
+    // modify the last hex byte of the merkle root
+    block.header.merkle_root = Hash("0x18cb29d05b784548fe385401df4e0a920d254" ~
+        "dd3c59ab8942888e40211acf05581ad8547a2c7d37d532a1465cdba0acaa2492f49" ~
+        "f30331cabebe2464c89e1554");
```

Comment:
You can store `Hash` and change it (it has `opSlice`) to make the diff simpler to spot

Reply:
Ah, I thought it was immutable.. cool.